staging/src/k8s.io/client-go/rest/request.go

```diff
@@ -869,6 +869,27 @@ func (r *Request) request(ctx context.Context, fn func(*http.Request, *http.Resp
 		req = req.WithContext(ctx)
 		req.Header = r.headers
 
+		// net/http2 retries a request if it receives a GOAWAY frame.
+		// GOAWAY frame has LastStreamID which represents the last stream identifier the server
+		// has processed or is aware of. On the client side, net/http2 will set an error for any
+		// stream that has an identifier higher than that of LastStreamID.
+		// However, if the request body is not empty then net/http2 will retry only when it can
+		// reset back the request body to its original state via the optional req.GetBody.
+		//
+		// This is where it happens in net/http2
+		// https://github.com/golang/go/blob/master/src/net/http/h2_bundle.go#L6991
+		//
+		// Let's set GetBody for our request so net/http2 automatically retry it. It's ideal if
+		// we allow net/http2 to retry a request before we do a retry here.
+		//
+		// net/http sets GetBody for some type of body but not all, so it's good to
+		// ensure that all requests with a body have a GetBody defined.
+		if r.body != nil && req.GetBody == nil {
```

Comment:
shouldn't this be limited to r.verb == GET or bodies that can be reset? the GetBody implemented here doesn't do anything to reset the body to its original state, which means it is not implemented according to the godoc for GetBody

Reply:
Thanks @liggitt for your feedback. 
The `Body` function sets the `body` of a `rest.Request` object to an instance of `*bytes.Reader` if the input is `Runtime.Object`, `[]byte` or `string`. On the other hand, If body is of type `*bytes.Reader` then `http.NewRequest` will set a `GetBody` for us - https://github.com/golang/go/blob/master/src/net/http/request.go#L881-L887

So I think we are covered, I have reduced my PR to a unit test that ensures that `rest.Request.body`, if not empty always produces a `http.Request` with a non-nil `GetBody`.